ui/raidboss/emulator/EmulatorCommon.ts

```diff
@@ -0,0 +1,127 @@
+          ret[i] = EmulatorCommon._cloneData(data[i]);
+
+        return ret;
+      }
+
+      if (data === null)
+        return null;
+
+      if (data instanceof RegExp)
+        return new RegExp(data);
+
+      const ret: DataType = {};
+      for (const i in data)
+        ret[i] = EmulatorCommon._cloneData(data[i]);
+
+      return ret;
+    }
+    return data;
+  }
+
+  static timeToString(time: number, includeMillis = true): string {
+    const negative = time < 0 ? '-' : '';
+    time = Math.abs(time);
+    const millisNum = time % 1000;
+    const secsNum = ((time % (60 * 1000)) - millisNum) / 1000;
+    // Milliseconds
+    const millis = `00${millisNum}`.substr(-3);
+    const secs = `0${secsNum}`.substr(-2);
+    const mins = `0${((((time % (60 * 60 * 1000)) - millisNum) / 1000) - secsNum) / 60}`.substr(-2);
+    return negative + mins + ':' + secs + (includeMillis ? '.' + millis : '');
+  }
+
+  static timeToDateString(time: number): string {
+    return this.dateObjectToDateString(new Date(time));
+  }
+
+  static dateObjectToDateString(date: Date): string {
+    const year = date.getFullYear();
+    const month = EmulatorCommon.zeroPad((date.getMonth() + 1).toString());
+    const day = EmulatorCommon.zeroPad(date.getDate().toString());
+    return `${year}-${month}-${day}`;
+  }
+
+  static timeToTimeString(time: number, includeMillis = false): string {
+    return this.dateObjectToTimeString(new Date(time), includeMillis);
+  }
+
+  static dateObjectToTimeString(date: Date, includeMillis = false): string {
+    const hour = EmulatorCommon.zeroPad(date.getHours().toString());
+    const minute = EmulatorCommon.zeroPad(date.getMinutes().toString());
+    const second = EmulatorCommon.zeroPad(date.getSeconds().toString());
+    let ret = `${hour}:${minute}:${second}`;
+    if (includeMillis)
+      ret = ret + `.${date.getMilliseconds()}`;
+
+    return ret;
+  }
+
+  static msToDuration(ms: number): string {
+    const tmp = EmulatorCommon.timeToString(ms, false);
+    return tmp.replace(':', 'm') + 's';
+  }
+
+  static dateTimeToString(time: number, includeMillis = false): string {
+    const date = new Date(time);
+    return `${this.dateObjectToDateString(date)} ${this.dateObjectToTimeString(date, includeMillis)}`;
+  }
+
+  static zeroPad(str: string, len = 2): string {
+    return ('' + str).padStart(len, '0');
+  }
+
+  static properCase(str: string): string {
+    return str.replace(/([^\W_]+[^\s-]*) */g, (txt) => {
+      return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
+    });
+  }
+
+  static spacePadLeft(str: string, len: number): string {
+    return str.padStart(len, ' ');
+  }
+
+  static sealRegexes: Record<string, RegExp> = LocaleNetRegex.areaSeal;
+  static engageRegexes: Record<string, RegExp> = LocaleNetRegex.countdownEngage;
+  static countdownRegexes: Record<string, RegExp> = LocaleNetRegex.countdownStart;
+  static unsealRegexes: Record<string, RegExp> = LocaleNetRegex.areaUnseal;
```

Comment:
The narrower type causes issues with `LogEventHandler`'s usage of a helper method to check against the records.

```js
  static doesLineMatch(line, regexes) {
    for (const i in regexes) {
      const res = regexes[i].exec(line);
      if (res) {
        if (LocaleNetRegex.areaSeal[i])
          res.groups.language = i;
        return res;
      }
    }
    return false;
  }
```

This is used for both the LocaleNetRegex exports as well as the other static RegExp values defined on EmulatorCommon, compare these two calls:
```js
    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.engageRegexes);
// vs
    res = LogEventHandler.doesLineMatch(line, [EmulatorCommon.winRegex]);
```

Reply:
Probably this can be resolved upstream in `LogEventHandler` by adjusting the `regexes` parameter to be a union type between the two possible values. Either way, I don't think this should hold up this PR, adjusting for next commit.